compressors/lz77.py

```diff
@@ -0,0 +1,705 @@
+"""
+LZ77 works by finding matches in the past and encoding the match lengths and offsets.
+Symbols not part of matches are directly stored as "literals". These streams are 
+then entropy coded. LZ77 forms the basis of popular compressors like gzip and zstd.
+Theoretically LZ77 is a universal compressor, that is, it asymptotically achieves the
+entropy rate of any stationary process.
+
+This is a simplified LZ77 encoder deriving ideas from: 
+- LZ77 modern implementation: https://glinscott.github.io/lz/index.html 
+- DEFLATE: https://www.rfc-editor.org/rfc/rfc1951 
+- ZSTD: https://github.com/facebook/zstd/blob/dev/doc/zstd_compression_format.md
+
+The encoder has 2 parameters: 
+- min_match_length: this determines the minimum
+allowed match length. The same length is also used for indexing into the dict.
+Very low values lead to slow execution. In terms of compression, low values can
+lead to suboptimal matches but high values can lead to missed matches. So there
+is a tradeoff. Default is 6.
+- max_num_matches_considered: Max number of matches considered at
+a position to bound time complexity. The matches are considered in order of
+recency and the longest match among the max_num_matches_considered is chosen.
+Defaults to 64.
+
+The algorithm: 
+
+- keep a window that stores everything the encoder has seen (not reset across blocks
+ unless reset called). 
+- keep a dict that stores a map from tuples of length of min_match_length to the 
+positions where the tuple is seen. 
+- to find a match during parsing, we look up in the dict and then find the longest
+match among the most recent max_num_matches_considered. If no match at a position, we
+search for matches starting at next. We do this until we reach the end of the block
+and don't find anything.
+- the streams generated by parsing are as follows:
+1. Literals - when no matches found for a position we put in literals. This can also include
+the very end of the block which is unmatched.
+2. LZ77 sequences - this is a triple of literal_counts, match_length and match_offset. The idea 
+is that when we start looking for a match starting at pos_in_window, but only find a match starting
+at match_start_pos, we set literal_counts to match_start_pos-pos_in_window. The match length and
+offset have their usual meaning. Note that as in gzip/zstd we allow match_length to be greater than
```

Comment:
maybe good to add what match_offset is `cur_pos - match_start_pos` (is that right?.. as it is not completely obvious)

Reply:
It's actually `match_start_pos-best_candidate_start_pos`, will to the description.